core/src/main/java/com/datastax/oss/driver/api/core/type/codec/TypeCodecs.java

```diff
@@ -163,9 +374,114 @@
    *
    * @see #ZONED_TIMESTAMP_SYSTEM
    * @see #ZONED_TIMESTAMP_UTC
+   * @see #ZONED_TIMESTAMP_PERSISTED
    */
   @NonNull
   public static TypeCodec<ZonedDateTime> zonedTimestampAt(@NonNull ZoneId timeZone) {
     return new ZonedTimestampCodec(timeZone);
   }
+
+  /**
+   * Returns a codec that handles Apache Cassandra(R)'s timestamp type and maps it to Java's {@link
+   * LocalDateTime}, using the supplied {@link ZoneId} as its source of time zone information.
+   *
+   * <p>Note that Apache Cassandra(R)'s timestamp type does not store any time zone; the codecs
+   * created by this method are provided merely as a convenience for users that need to deal with
+   * local date-times in their applications.
+   *
+   * @see #LOCAL_TIMESTAMP_UTC
+   * @see #localTimestampAt(ZoneId)
+   */
+  @NonNull
+  public static TypeCodec<LocalDateTime> localTimestampAt(@NonNull ZoneId timeZone) {
+    return new LocalTimestampCodec(timeZone);
+  }
+
+  /**
+   * Returns a codec mapping CQL lists to Java object arrays. Serialization and deserialization of
+   * elements in the array is delegated to the provided element codec.
+   *
+   * <p>This method is not suitable for Java primitive arrays. Use {@link #BOOLEAN_ARRAY}, {@link
+   * #BYTE_ARRAY}, {@link #SHORT_ARRAY}, {@link #INT_ARRAY}, {@link #LONG_ARRAY}, {@link
+   * #FLOAT_ARRAY} or {@link #DOUBLE_ARRAY} instead.
+   */
+  @NonNull
+  public static <T> TypeCodec<T[]> arrayOf(@NonNull TypeCodec<T> elementCodec) {
+    return new ObjectArrayCodec<>(elementCodec);
+  }
+
+  /**
+   * Returns a codec mapping Java Enums to CQL ints, according to their {@linkplain Enum#ordinal()
+   * ordinals}.
+   *
+   * @see #enumNamesOf(Class)
+   */
+  @NonNull
+  public static <EnumT extends Enum<EnumT>> TypeCodec<EnumT> enumOrdinalsOf(
+      @NonNull Class<EnumT> enumClass) {
+    return new EnumOrdinalCodec<>(enumClass);
+  }
```

Comment:
I wonder if we should provide this out of the box. Using ordinals is a terrible idea, let's not encourage it.

Reply:
But we already provided this in driver 3.x, do you think we should remove this functionality then?

Reply:
I'm in favor of leaving it. Ordinals can save space on disk, if the enum never changes I don't see why it's so horrible.

Reply:
Sorry to rekindle the debate, I was about to concede on the grounds of 3.x compatibility, but this really bothers me:

> if the enum never changes I don't see why it's so horrible

That's a very big if. The enum always changes. It just takes one careless developer inserting a new constant in the middle, or reordering the constants to their liking, and you start inserting different codes, that overlap with the old ones. Your database is trashed, and it's not recoverable (beyond restoring a backup, or maybe a script that would track the writetime, but imagine the mess).

It's far from evident: the enum might come from a different layer of the application, or even a different project, where people are not even aware that it's being inserted in a database.

It's almost impossible to enforce: you can add a comment ("the order of these constants matters"), but you don't know if it will be read. Tools like Clirr or Revapi will warn you, but I'm not sure how widespread their use is.

So all in all it's a very dangerous feature, I think we are doing our users a disservice by providing it. It was indeed in 3.x, but I think it was a mistake to include it then.
If people want to save bytes, the better solution is to have an explicit `fromCode` / `toCode` conversion layer, and write your own codec, which should be relatively straightforward with `MappingCodec`.

Reply:
>  It just takes one careless developer inserting a new constant in the middle, or reordering the constants to their liking

... or changing an enum constant name. And your database is screwed just the same. In this case we shouldn't provide any codec for enums. And then we would be back to careful users complaining that our driver is hard to use because they need to create their codecs themselves.

There is demand for storing enums in the database. Someone in DataStax Community recently was even [asking for an enum type in Cassandra itself](https://community.datastax.com/questions/5844/enum-datatype-for-cassandra-column.html). One of the reasons invoked is storage efficiency, and enum ordinals would be better at that.

I would like us to provide both codecs, just like in driver 3. We are not responsible for careless developers, we have warnings in the docs explaining the caveats of both approaches, so we're covered imo.

Reply:
> or changing an enum constant name. And your database is screwed just the same

No, it's much more obvious: client code will stop compiling, and even after that the codec will fail to decode the old name. Also, you would likely pick an entirely new name, not swap existing constants.

Let's say you store card suits by name, but one of the names is wrong: `enum Suit { HEARTS, DIAMONDS, SPADES, CLOVERS }`. Someone fixes `CLOVERS` => `CLUBS`. Assuming no one picked up on the compile errors, and your tests didn't catch it either, the worst that can happen is that you only find out in production. This will fail hard and fast as soon as the codec tries to decode `CLOVERS`. You revert the upgrade, and you also need an annoying but doable migration of any corrupted writes: find any row with `CLUBS` and replace it with `CLOVERS`.

On the other hand, you store suits by ordinal, and have `enum Suit { HEARTS, DIAMONDS, SPADES, CLUBS }`. But someone decides that enum constants should be sorted in alphabetical order: `enum Suit { CLUBS, DIAMONDS, HEARTS, SPADES }`. The code still compiles. Tests won't catch it if they use enum constants for both reads & writes. Production code won't fail fast, it still happily reads and writes data. It's more likely that a human will detect the issue, which might take significantly longer. When all is said and done, you're left to reconcile rows where 0 means hearts, 2 means spades and 3 means clubs, with rows where 0 means clubs, 2 means hearts and 3 means spades. This is an order of magnitude worse.

> We are not responsible for careless developers

No, but we shouldn't give users a ticking timebomb in a nice giftwrap either.

Reply:
@csplinter any opinion on this?
I'm not dead-set on not having an enum ordinal codec out of the box, but there are strong arguments against it.